README
= flickr-fu

== Contact

  Author: Ben Wyrosdick
  Email: ben [at] commonthread.com
  Lighthouse: http://commonthread.lighthouseapp.com/projects/12069-flickr_fu/overview
  Main Repository: http://github.com/commonthread/flickr_fu/tree/master

== Getting Started

  You need to first get an API key as detailed here:

  http://www.flickr.com/services/api/misc.api_keys.html
  
== Installation

  sudo gem install flickr-fu

== Documentation

  RDoc Documentation can be found here:

  http://www.commonthread.com/projects/flickr_fu/rdoc/

== Example flickr.yml
--- !map:HashWithIndifferentAccess
key: "YOUR KEY"
secret: "YOUR SECRET"
token_cache: "token_cache.yml"

== Authorization

  To authorise your application to access Flickr using your API key you will
  need to access a specific URL.
  
  To generate this URL run the following and when presented with the URL 
  access it from your browser. Confirm the application has permission at 
  the level you have specified.
  
  Note that flickr has different types of keys. If you use one for a webapplication,
  which is most likely, you need to define a callback and somehow make sure that
  you connect the parameter :frob that flickr send via the callback is assigned
  to the right user account. The best way to do this is to loop over all the current
  user's flickr accounts and try flickr.auth.token with the :frob.
  
  Finally, cache the token (this will create the token cache file)
  
  If you have an invalid API key you will see errors such as:
  
  "100: Invalid API Key"
  
  If you don't follow the process below to authorise your application
  you will see errors such as:
  
  "98: Login failed / Invalid auth token" or 
  "99: User not logged in / Insufficient permissions"
  
== Authorization Example for non-webapplication

  require 'flickr_fu'
  
  flickr = Flickr.new('flickr.yml')
  
  puts "visit the following url, then click <enter> once you have authorized:"

  # request write permissions
  puts flickr.auth.url(:write)
  
  gets
  
  flickr.auth.cache_token

== Authorization Example for a webapplication

  flickr.auth.token also contains the nsid and username, this
  example only stores the token and no other userdata.
  All you need in the views is a button or link to FlickrController.create

  require 'flickr_fu'
  class FlickrController < ActionController::Base
    def create
      flickr = Flickr.new('flickr.yml')
      redirect_to flickr.auth.url(:write)
    end
    def flickr_callback
      flickr = Flickr.new('flickr.yml')
      flickr.auth.frob = params[:frob]
      current_user.update_attribute :flickr_token, flickr.auth.token.token
    end
    def something_else_with_flickr
      flickr = Flickr.new(YAML.load_file('flickr.yml').merge(:token => current_user.flickr_token))
      # now you have full access on the user's data :)
    end
  end

=== User creation on the fly

  To use the userdata in flickr.auth.token and create
  an account with it on the fly try following flickr_callback
  
  def flickr_callback
    flickr = Flickr.new('flickr.yml')
    flickr.auth.frob = params[:frob]
    if(!logged_in?)
      # no user yet for this flickr id, create a new user
      if (current_user = User.find_by_flickr_nsid(flickr.auth.token.user_id))).nil?
        current_user = User.new do |user|
          user.login = flickr.auth.token.username
          user.real_name = flickr.auth.token.user_real_name
          user.flickr_nsid = flickr.auth.token.user_id
          user.flickr_token = flickr.auth.token.token
        end
      else # we have a user for that flick account, login and update token
        current_user.flickr_token = flickr.auth.token.token
      end
      if(current_user.save)
        flash[:notice] = 'Registered and logged in via flickr'
        redirect_to '/'
      else
        // probably you validate for the presence of an email address?
        flash[:error] = 'Logged in via flickr but some user data is missing'
        @user = current_user
        render :action => 'users/new'
      end
    else
      current_user.update_attribute :flickr_token, flickr.auth.token.token
    end
  end

== Search Example

  require 'flickr_fu'

  flickr = Flickr.new('flickr.yml')

  photos = flickr.photos.search(:tags => 'ruby-flickr')

  puts "found #{photos.size} photo(s)"

  photos.each do |photo|
    puts photo.title
    puts photo.description unless [nil, ''].include?(photo.description)
    [:square, :thumbnail, :small, :medium, :large, :original].each do |size|
      puts "#{size}: #{photo.url(size)}"
    end
    puts "comments: #{photo.comments.size}"
    photo.comments.each do |comment|
      intro = "#{comment.author_name} says - "
      puts "#{intro}\"#{comment.comment.gsub("\n", "\n"+(" "*intro.length))}\""
    end
    puts "notes: #{photo.notes.size}"
    photo.notes.each do |note|
      puts "[#{note.x},#{note.y} ~ #{note.width}x#{note.height}] - \"#{note.note}\""
    end
    puts
    puts
  end
  
== Another Search Example

  If searching for photos by user id then you need to specify the 'alias' - without
  intervention this is usually set by Flickr and is an alphanumeric string. 
  
  To find out the user id for a given user, you can use the tool at:
  
  http://idgettr.com/
  
  And replace the line in the above sample to query on user id:
  
  photos = flickr.photos.search(:user_id => 'your_user_id_here')

== Authenticatd Search Example

  For a few searches you will need to authenticate. E.g. to receive the original_url
  staight away (and thus saving an extra getInfo call), or to search for the private
  photos of a user. In order to do so you need to pass the user's token and his
  flickr_nsid (though the API docs also suggest 'me', both work fine).

  This example appends the :url_o to the extras, mapping it to :original_url:

  flickr.photos.extras << {:url_o => :original_url}
  photos = flickr.photos.search(:per_page => 5, :page => 1,
      :user_id => 'me', :privacy_filter => 5, :auth_token => user_token,
      :safe_search => 3))

== Patch Contributers

  Chris Ledet
  Maciej Biłas
  Mike Perham
  Chris Anderton
  Luke Francl
  Thomas R. Koll
  P. Mark Anderson
  Josh Nichols


CHANGELOG
2010-07-07: 0.3.4 (tomk32-flickr_fu)
  * Added medium640 size (the new 'z' from flickr's brand-new broader layout)
  * Added attributes for height and width of original

2010-04-22: 0.3.3 (tomk32-flickr_fu)
  * Fixed bug where :extra fields received via photos.search where ignored in Photo
    and an additional API call was made
2010-04-17: 0.3.1 (tomk32-flickr_fu)
  * Allows to add extra fields to flickr.photos.search (and other in this flickr.photo)
    that are requested from flickr. This can be used to e.g. get the original_url quicker.


lib/flickr/base.rb
require 'yaml'
module Flickr
  def self.new(*params)
    Flickr::Base.new(*params)
  end
  
  class Base
    attr_reader :api_key, :api_secret, :token_cache, :token
    
    REST_ENDPOINT = 'http://api.flickr.com/services/rest/'
    AUTH_ENDPOINT = 'http://flickr.com/services/auth/'
    UPLOAD_ENDPOINT = 'http://api.flickr.com/services/upload/'
    
    # create a new flickr object
    # 
    # You can either pass a hash with the following attributes:
    # 
    # * :key (Required)
    #     the API key
    # * :secret (Required)
    #     the API secret
    # * :token (Optional)
    #     Flickr::Auth::Token object
    #
    # or:
    # * config_file (Required)
    #     yaml file to load configuration from
    # * token_cache (Optional)
    #     location of the token cache file. This will override the setting in the config file
    # 
    # Config Example (yaml file)
    # ---
    # key: YOUR_API_KEY
    # secret: YOUR_API_SECRET
    # token_cache: token.yml
    # 
    def initialize(config_hash_or_file, token_cache = nil)
      if config_hash_or_file.is_a? Hash
        @api_key = config_hash_or_file[:key]
        @api_secret = config_hash_or_file[:secret]
        @token = config_hash_or_file[:token]
        raise 'config_hash must contain api key and secret' unless @api_key and @api_secret
      else 
        config = YAML.load_file(config_hash_or_file)
      
        @api_key = config['key']
        @api_secret = config['secret']
        @token_cache = token_cache || config['token_cache']
        raise 'flickr config file must contain an api key and secret' unless @api_key and @api_secret
      end
    end

    # sends a request to the flickr REST api
    # 
    # Params
    # * method (Required)
    #     name of the flickr method (ex. flickr.photos.search)
    # * options (Optional)
    #     hash of query parameters, you do not need to include api_key, api_sig or auth_token because these are added automatically
    # * http_method (Optional)
    #     choose between a GET and POST http request. Valid options are:
    #       :get (DEFAULT)
    #       :post
    # * endpoint (Optional)
    #     url of the api endpoint
    # 
    def send_request(method, options = {}, http_method = :get, endpoint = REST_ENDPOINT)
      options.merge!(:api_key => @api_key, :method => method)
      sign_request(options)
      
      if http_method == :get
        api_call = endpoint + "?" + options.collect{|k,v| "#{k}=#{CGI.escape(v.to_s)}"}.join('&')
        rsp = Net::HTTP.get(URI.parse(api_call))
      else
        rsp = Net::HTTP.post_form(URI.parse(endpoint), options).body
      end
      
      rsp = '<rsp stat="ok"></rsp>' if rsp == ""
      xm = XmlMagic.new(rsp)
      
      if xm[:stat] == 'ok'
        xm
      elsif Errors::ERROR_CODES.include?(xm.err[:code].to_i)
        raise Errors::ERROR_CODES[xm.err[:code].to_i].new, "#{xm.err[:code]}: #{xm.err[:msg]}"
      else
        raise Errors::UnknownError.new, "#{xm.err[:code]}: #{xm.err[:msg]}"
      end
    end
    
    # alters your api parameters to include a signiture and authorization token
    # 
    # Params
    # * options (Required)
    #     the hash of parameters to be passed to the send_request
    # * authorize (Optional)
    #     boolean value to determine if the call with include an auth_token (Defaults to true)
    # 
    def sign_request(options, authorize = true)
      options.merge!(:auth_token => self.auth.token(false).to_s, :api_key => @api_key) if authorize and self.auth.token(false)
      options.delete(:api_sig)
      options.merge!(:api_sig => Digest::MD5.hexdigest(@api_secret + options.to_a.sort_by{|k| k[0].to_s}.flatten.join)) if @api_secret
    end
    
    # creates and/or returns the Flickr::Test object
    def test() @test ||= Flickr::Test.new(self) end

    # creates and/or returns the Flickr::Photos object
    def photos() @photos ||= Flickr::Photos.new(self) end
      
    # creates and/or returns the Flickr::People object
    def people() @people ||= Flickr::People.new(self) end
      
    # creates and/or returns the Flickr::Auth object
    def auth() @auth ||= Flickr::Auth.new(self) end
      
    # creates and/or returns the Flickr::Uploader object
    def uploader() @uploader ||= Flickr::Uploader.new(self) end
  end
end

lib/flickr/photo.rb
# wrapping class to hold an flickr photo
class Flickr::Photos::Photo
  attr_accessor :id, :owner, :secret, :server, :farm, :title, :is_public, :is_friend, :is_family # standard attributes
  attr_accessor :license_id, :uploaded_at, :taken_at, :owner_name, :icon_server, :original_format, :updated_at, :geo, :tags, :machine_tags, :o_dims, :views, :media # extra attributes
  attr_accessor :comments # comment attributes
  
  # create a new instance of a flickr photo.
  # 
  # Params
  # * flickr (Required)
  #     the flickr object
  # * attributes (Required)
  #     a hash of attributes used to set the initial values of the photo object
  def initialize(flickr, attributes)
    @flickr = flickr
    attributes.each do |k,v|
      # necessary for anything that comes via 'extras' or of course
      # possible future changes in the api returns
      if ! respond_to?(k)
        self.class.send(:attr_accessor, k)
      end
      send("#{k}=", v)
    end
  end

  # returns an instance of Flickr::Photos::Size for the required size
  #
  # Params
  #  * size (Optional)
  #    the size of the size instance to return.  Optional sizes are:
  #       :square - square 75x75
  #       :thumbnail - 100 on longest side
  #       :small - 240 on longest side
  #       :medium - 500 on longest side
  #       :large - 1024 on longest side (only exists for very large original images)
  #       :original - original image, either a jpg, gif or png, depending on source format
  # Examples
  #       Photo.photo_size(:square).source
  #       Photo.photo_size(:large).width
  def photo_size(size = :medium)
    size_hash.fetch(size.to_s, size_hash['medium'])
  end

  # retreive the url to the image stored on flickr
  #
  # == Params
  # * size (Optional)
  #     the size of the image to return. Optional sizes are:
  #       :square - square 75x75
  #       :thumbnail - 100 on longest side
  #       :small - 240 on longest side
  #       :medium - 500 on longest side
  #       :large - 1024 on longest side (only exists for very large original images)
  #       :original - original image, either a jpg, gif or png, depending on source format
  #
  def image_url(size = :medium)
    # It turns out that flickr always stores all the sizes of the picture even when getSizes call returns otherwise.
    # Not calling getSizes is also very important for performance reasons.
    # Retrieving 30 search results means calling the API 31 times if you call getSizes every time.
    # Mind that you still need to call getSizes if you go out for the original image.
    if size == :original
      return original_url if respond_to?(:original_url) and !original_url.blank?
      size_hash[size.to_s].source if size_hash.has_key? size.to_s
    else
      key = "_#{size_key(size.to_sym)}"
      key = "" if key == "_"
      "http://farm#{farm}.static.flickr.com/#{server}/#{id}_#{secret}#{key}.jpg"
    end
  end
  alias :url :image_url

  def photopage_url
    # Keeping the same convention as image_url (foo_url)
    url_photopage
  end

  def video_url
    if size_hash['video player']
      size_hash['video player'].source
    end
  end

  # save the current photo to the local computer
  # 
  # == Params
  # * filename (Required)
  #     name of the new file omiting the extention (ex. photo_1)
  # * size (Optional)
  #     the size of the image to return. Optional sizes are:
  #       :small - square 75x75
  #       :thumbnail  - 100 on longest side
  #       :small      - 240 on longest side
  #       :medium     - 500 on longest side
  #       :medium_500 - 500 on the longest side
  #       :medium_640 - 640 on the longest side
  #       :large      - 1024 on longest side (only exists for very large original images)
  #       :original   - original image, either a jpg, gif or png, depending on source format
  # 
  def save_as(filename, size = :medium)
    format = size.to_sym == :original ? (self.original_format || 'jpg') : 'jpg'
    filename = "#{filename}.#{format}"

    if File.exists?(filename) or not self.url(size)
      false
    else
      f = File.new(filename, 'w+')
      f.puts open(self.url(size)).read
      f.close
      f
    end
  end
  
  # Add tags to a photo.
  # 
  # Params
  # * tags (Required)
  #     comma seperated list of tags
  # 
  def add_tags(tags)
    @flickr.send_request('flickr.photos.addTags', {:photo_id => self.id, :tags => tags}, :post)
    true
  end
  
  # Add comment to a photo as the currently authenticated user.
  #
  # Params
  # * message (Required)
  #     text of the comment
  #
  def add_comment(message)
    @flickr.send_request('flickr.photos.comments.addComment', {:photo_id => self.id, :comment_text => message}, :post)
    true
  end
  
  # Add a note to a photo. Coordinates and sizes are in pixels, based on the 500px image size shown on individual photo pages.
  # 
  # Params
  # * message (Required)
  #     The text of the note
  # * x (Required)
  #     The left coordinate of the note
  # * y (Required)
  #     The top coordinate of the note
  # * w (Required)
  #     The width of the note
  # * h (Required)
  #     The height of the note
  #     
  def add_note(message, x, y, w, h)
    @flickr.send_request('flickr.photos.notes.add', {:photo_id => self.id, :note_x => x, :note_y => y, :note_w => w, :note_h => h, :note_text => message}, :post)
    true
  end
  
  # Rotate a photo.
  # 
  # Params
  # * degrees (Required)
  #     The amount of degrees by which to rotate the photo (clockwise) from it's current orientation. Valid values are 90, 180 and 270.
  # 
  def rotate(degrees)
    @flickr.send_request('flickr.photos.transform.rotate', {:photo_id => self.id, :degrees => degrees}, :post)
    true
  end
  
  # return the license associated with the photo
  # 
  def license
    @flickr.photos.licenses[self.license_id]
  end

  # Returns the location of the photo (if available)
  # or nil if photo is not geo-tagged.
  def location
    begin
      @location ||= @flickr.photos.geo.get_location(self.id)
    rescue Flickr::Error => e
      if e.code == 2 # 2: Photo has no location information.
        return nil
      else
        raise e
      end
    end
  end

  def location= location
    if !location.nil?
      @flickr.photos.geo.set_location(self.id, location.latitude, location.longitude, location.accuracy)
    else
      @flickr.photos.geo.remove_location(self.id)
    end
    @location = location
  end
  
  # Sets the license for a photo.
  # 
  # Params
  # * license_id (Required)
  #     The license to apply, or 0 (zero) to remove the current license.
  def set_license(license_id)
    @flickr.send_request('flickr.photos.licenses.setLicense', {:photo_id => self.id, :license_id => license_id}, :post)
    true
  end

  def public?
    is_public == "1"
  end

  def friend?
    is_friend == "1"
  end

  def family?
    is_family == "1"
  end

  def comments # :nodoc:
    @comments ||= begin
      if @comment_count == 0
        self.comments = []
      else
        rsp = @flickr.send_request('flickr.photos.comments.getList', :photo_id => self.id)
        
        self.comments = []
        
        rsp.comments.comment.each do |comment|
          self.comments << Flickr::Photos::Comment.new(:id => comment[:id],
            :comment => comment.to_s,
            :author => comment[:author],
            :author_name => comment[:authorname],
            :permalink => comment[:permalink],
            :created_at => (Time.at(comment[:datecreate].to_i) rescue nil))
        end
      end

      self.comments
    end
  end
  
  def sizes # :nodoc:
    @sizes ||= begin
      rsp = @flickr.send_request('flickr.photos.getSizes', :photo_id => self.id)
      
      _sizes = []
      rsp.sizes.size.each do |size|
        _sizes << Flickr::Photos::Size.new(:label => size[:label], :width => size[:width],
          :height => size[:height], :source => size[:source], :url => size[:url])
      end
      _sizes
    end
  end

  def method_missing(method, *args, &block)
    if [:description, :original_secret, :owner_username, :owner_realname, :url_photopage, :notes].include?(method.to_sym)
      @attach_info ||= attach_info
      return @attach_info[method.to_sym]
    end
  end

  protected
  def size_hash
    @size_hash ||= begin
      hash = {}
      sizes.each do |size|
        hash[size.label.downcase] = size
      end
      hash
    end
  end

  private
  attr_accessor :comment_count
  
  # convert the size to the key used in the flickr url
  def size_key(size)
    case size.to_sym
    when :square then 's'
    when :thumb, :thumbnail then 't'
    when :small then 'm'
    when :medium then ''
    when :medium_500 then ''
    when :medium_640 then 'z'
    when :large then 'b'
    when :original then 'o'
    else ''
    end
  end

  # loads photo info when a field is requested that requires additional info
  def attach_info
    rsp = @flickr.send_request('flickr.photos.getInfo', :photo_id => self.id, :secret => self.secret)

    if rsp.photo.notes.note
      notes = rsp.photo.notes.note.collect do |note|
        Flickr::Photos::Note.new(:id => note[:id],
          :note => note.to_s,
          :author => note[:author],
          :author_name => note[:authorname],
          :x => note[:x],
          :y => note[:y],
          :width => note[:w],
          :height => note[:h])
      end
    end
    return {
      :description => rsp.photo.description.to_s.strip,
      :original_secret => rsp.photo[:originalsecret],
      :owner_username => rsp.photo.owner[:username],
      :owner_realname => rsp.photo.owner[:realname],
      :url_photopage => rsp.photo.urls.url.to_s,
      :comment_count => rsp.photo.comments.to_s.to_i,
      :notes => notes
    }
  end
end


lib/flickr/photos.rb
class Flickr::Photos < Flickr::Base
  attr_accessor :extras
  def initialize(flickr)
    @flickr = flickr
    self.extras ||= {
      # flickr => flickr_fu
      :license => :license_id,
      :date_upload => :date_upload,
      :datetaken => :taken_at,
      :ownername => :owner_name,
      :icon_server => :icon_server,
      :originalformat => :original_format,
      :lastupdate => :updated_at,
      :machine_tags => :machine_tags,
      :geo => :geo,
      :tags => :tags,
      :o_dims => :o_dims,
      :views => :views,
      :media => :media
    }
  end

  # Return a list of photos matching some criteria. Only photos visible to the calling user will be returned. To return private or semi-private photos,
  # the caller must be authenticated with 'read' permissions (by passing :auth_token), and have permission to view the photos. Unauthenticated calls will only return public photos.
  #
  # == Authentication
  # This method does not require authentication.
  #
  # == Options
  # * user_id (Optional)
  #     The NSID of the user who's photo to search. If this parameter isn't passed then everybody's public photos will be searched. A value of "me" will
  #     search against the calling user's photos for authenticated calls.
  # * tags (Optional)
  #     A comma-delimited list of tags. Photos with one or more of the tags listed will be returned.
  # * tag_mode (Optional)
  #     Either 'any' for an OR combination of tags, or 'all' for an AND combination. Defaults to 'any' if not specified.
  # * text (Optional)
  #     A free text search. Photos who's title, description or tags contain the text will be returned.
  # * min_upload_date (Optional)
  #     Minimum upload date. Photos with an upload date greater than or equal to this value will be returned. The date should be in the form of a unix timestamp.
  # * max_upload_date (Optional)
  #     Maximum upload date. Photos with an upload date less than or equal to this value will be returned. The date should be in the form of a unix timestamp.
  # * min_taken_date (Optional)
  #     Minimum taken date. Photos with an taken date greater than or equal to this value will be returned. The date should be in the form of a mysql datetime.
  # * max_taken_date (Optional)
  #     Maximum taken date. Photos with an taken date less than or equal to this value will be returned. The date should be in the form of a mysql datetime.
  # * license (Optional)
  #     The license id for photos (for possible values see the flickr.photos.licenses.getInfo method). Multiple licenses may be comma-separated.
  # * sort (Optional)
  #     The order in which to sort returned photos. Deafults to date-posted-desc. The possible values are: date-posted-asc, date-posted-desc, date-taken-asc,
  #     date-taken-desc, interestingness-desc, interestingness-asc, and relevance.
  # * privacy_filter (Optional)
  #     Return photos only matching a certain privacy level. This only applies when making an authenticated call to view photos you own. Valid values are:
  #       1 public photos
  #       2 private photos visible to friends
  #       3 private photos visible to family
  #       4 private photos visible to friends & family
  #       5 completely private photos
  # * bbox (Optional)
  #     A comma-delimited list of 4 values defining the Bounding Box of the area that will be searched.
  #
  #     The 4 values represent the bottom-left corner of the box and the top-right corner, minimum_longitude, minimum_latitude, maximum_longitude, maximum_latitude.
  #
  #     Longitude has a range of -180 to 180 , latitude of -90 to 90. Defaults to -180, -90, 180, 90 if not specified.
  #
  #     Unlike standard photo queries, geo (or bounding box) queries will only return 250 results per page.
  #
  #     Geo queries require some sort of limiting agent in order to prevent the database from crying. This is basically like the check against "parameterless searches"
  #     for queries without a geo component.
  #
  #     A tag, for instance, is considered a limiting agent as are user defined min_date_taken and min_date_upload parameters — If no limiting factor is passed we
  #     return only photos added in the last 12 hours (though we may extend the limit in the future).
  # * accuracy (Optional)
  #     Recorded accuracy level of the location information. Current range is 1-16 :
  #       World level is 1
  #       Country is ~3
  #       Region is ~6
  #       City is ~11
  #       Street is ~16
  #     Defaults to maximum value if not specified.
  # * safe_search (Optional)
  #     Safe search setting:
  #       1 for safe.
  #       2 for moderate.
  #       3 for restricted.
  #     (Please note: Un-authed calls can only see Safe content.)
  # * content_type (Optional)
  #     Content Type setting:
  #       1 for photos only.
  #       2 for screenshots only.
  #       3 for 'other' only.
  #       4 for photos and screenshots.
  #       5 for screenshots and 'other'.
  #       6 for photos and 'other'.
  #       7 for photos, screenshots, and 'other' (all).
  # * machine_tags (Optional)
  #     Aside from passing in a fully formed machine tag, there is a special syntax for searching on specific properties :
  #       Find photos using the 'dc' namespace : "machine_tags" => "dc:"
  #       Find photos with a title in the 'dc' namespace : "machine_tags" => "dc:title="
  #       Find photos titled "mr. camera" in the 'dc' namespace : "machine_tags" => "dc:title=\"mr. camera\"
  #       Find photos whose value is "mr. camera" : "machine_tags" => "*:*=\"mr. camera\""
  #       Find photos that have a title, in any namespace : "machine_tags" => "*:title="
  #       Find photos that have a title, in any namespace, whose value is "mr. camera" : "machine_tags" => "*:title=\"mr. camera\""
  #       Find photos, in the 'dc' namespace whose value is "mr. camera" : "machine_tags" => "dc:*=\"mr. camera\""
  #     Multiple machine tags may be queried by passing a comma-separated list. The number of machine tags you can pass in a single query depends on
  #     the tag mode (AND or OR) that you are querying with. "AND" queries are limited to (16) machine tags. "OR" queries are limited to (8).
  # * machine_tag_mode (Required)
  #     Either 'any' for an OR combination of tags, or 'all' for an AND combination. Defaults to 'any' if not specified.
  # * group_id (Optional)
  #     The id of a group who's pool to search. If specified, only matching photos posted to the group's pool will be returned.
  # * woe_id (Optional)
  #     A 32-bit identifier that uniquely represents spatial entities. (not used if bbox argument is present). Experimental.
  #
  #     Geo queries require some sort of limiting agent in order to prevent the database from crying. This is basically like the check against "parameterless searches"
  #     for queries without a geo component.
  #
  #     A tag, for instance, is considered a limiting agent as are user defined min_date_taken and min_date_upload parameters &emdash; If no limiting factor is passed
  #     we return only photos added in the last 12 hours (though we may extend the limit in the future).
  # * place_id (Optional)
  #     A Flickr place id. (not used if bbox argument is present). Experimental.
  #
  #     Geo queries require some sort of limiting agent in order to prevent the database from crying. This is basically like the check against "parameterless searches"
  #     for queries without a geo component.
  #
  #     A tag, for instance, is considered a limiting agent as are user defined min_date_taken and min_date_upload parameters &emdash; If no limiting factor is passed
  #     we return only photos added in the last 12 hours (though we may extend the limit in the future).
  # * per_page (Optional)
  #     Number of photos to return per page. If this argument is omitted, it defaults to 100. The maximum allowed value is 500.
  # * page (Optional)
  #     The page of results to return. If this argument is omitted, it defaults to 1.
  # * media (Optional)
  #     The type of media to search for. 'photo', 'video', or 'both' are allowed arguments, with 'both' being the default.
  # * extras (Optional)
  #     Any extra fields of information you want. The default extra fields are defined
  #     in Flickr::Photos.extras and can be overwritten
  #
  def search(options)
    options[:extras] ||= self.extras.keys.join(',')

    rsp = @flickr.send_request('flickr.photos.search', options)

    returning PhotoResponse.new(:page => rsp.photos[:page].to_i,
                                :pages => rsp.photos[:pages].to_i,
                                :per_page => rsp.photos[:perpage].to_i,
                                :total => rsp.photos[:total].to_i,
                                :photos => [],
                                :api => self,
                                :method => 'search',
                                :options => options) do |photos|
      rsp.photos.photo.each do |photo|
        attributes = create_attributes(photo)

        photos << Photo.new(@flickr, attributes)
      end if rsp.photos.photo
    end
  end

  # Returns a list of the latest public photos uploaded to flickr.
  #
  # == Authentication
  # This method does not require authentication.
  #
  # == Options
  # * per_page (Optional)
  #     Number of photos to return per page. If this argument is omitted, it defaults to 100. The maximum allowed value is 500.
  # * page (Optional)
  #     The page of results to return. If this argument is omitted, it defaults to 1.
  # * media (Optional)
  #     The type of media to search for. 'photo', 'video', or 'both' are allowed arguments, with 'both' being the default.
  #
  def get_recent(options = {})
    options[:extras] ||= self.extras.keys.join(',')

    rsp = @flickr.send_request('flickr.photos.getRecent', options)

    returning PhotoResponse.new(:page => rsp.photos[:page].to_i,
                                :pages => rsp.photos[:pages].to_i,
                                :per_page => rsp.photos[:perpage].to_i,
                                :total => rsp.photos[:total].to_i,
                                :photos => [], :api => self,
                                :method => 'flickr.photos.getRecent',
                                :options => options) do |photos|
      rsp.photos.photo.each do |photo|
        attributes = create_attributes(photo)

        photos << Photo.new(@flickr, attributes)
      end if rsp.photos.photo
    end
  end

  def interesting(options)
    options[:extras] ||= self.extras.keys.join(',')

    rsp = @flickr.send_request('flickr.interestingness.getList', options)

    returning PhotoResponse.new(:page => rsp.photos[:page].to_i,
                                :pages => rsp.photos[:pages].to_i,
                                :per_page => rsp.photos[:perpage].to_i,
                                :total => rsp.photos[:total].to_i,
                                :photos => [],
                                :api => self,
                                :method => 'flickr.interestingness.getList',
                                :options => options) do |photos|
      rsp.photos.photo.each do |photo|
        attributes = create_attributes(photo)


        photos << Photo.new(@flickr, attributes)
      end if rsp.photos.photo
    end
  end

  def licenses
    @licenses ||= begin
      rsp = @flickr.send_request('flickr.photos.licenses.getInfo')

      returning Hash.new do |licenses|
        rsp.licenses.license.each do |license|
          licenses[license[:id].to_i] = Flickr::Photos::License.new(:id => license[:id].to_i, :name => license[:name], :url => license[:url])
        end
      end
    end
  end

  # creates and/or returns the Flickr::Photos::Geo object
  def geo
    @geo ||= Flickr::Photos::Geo.new(@flickr)
  end

  # Returns a Flickr::Photos::Photo object of the given id.
  # Raises an error if photo not found
  def find_by_id(photo_id)
    rsp = @flickr.send_request('flickr.photos.getInfo', :photo_id => photo_id)
    Photo.new(@flickr, :id => rsp.photo[:id].to_i, :owner => rsp.photo.owner,
      :secret => rsp.photo[:secret], :server => rsp.photo[:server].to_i, :farm => rsp.photo[:farm],
      :title => rsp.photo.title,
      :is_public => rsp.photo.visibility[:public], :is_friend => rsp.photo.visibility[:is_friend], :is_family => rsp.photo.visibility[:is_family])
  end

  protected
  def create_attributes(photo)
    attributes = {}

    # map all attributes from flickr to `attributes`
    {
      # flickr -> flickr_fu
      :id => :id,
      :owner => :owner,
      :secret => :secret,
      :server => :server,
      :farm => :farm,
      :title => :title,
      :ispublic => :is_public,
      :isfriend => :is_friend,
      :isfamily => :is_family,
      :o_width => :width,
      :o_height => :height
    }.merge(self.extras).each do |flickr_key, photo_key|
      attributes[photo_key] = photo[flickr_key]
    end

    # and a few conversions
    attributes.merge({
     :license_id => photo[:license].to_i,
     :uploaded_at => (Time.at(photo[:dateupload].to_i) rescue nil),
     :taken_at => (Time.parse(photo[:datetaken]) rescue nil),
     :updated_at => (Time.at(photo[:lastupdate].to_i) rescue nil),
     :views => photo[:views].to_i
    })
  end
end
